src/balance/helpers.rs

```diff
@@ -0,0 +1,354 @@
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+
+pub trait BTreeMapCoinHelpers {
+    fn into_vec(self) -> Vec<Coin>;
+    fn inplace_sub<'a, I>(&mut self, balance: I) -> StdResult<()>
+    where
+        I: IntoIterator<Item = (&'a String, &'a Uint128)>;
+
+    fn inplace_add<'a, I>(&mut self, balance: I) -> StdResult<()>
+    where
+        I: IntoIterator<Item = (&'a String, &'a Uint128)>;
+}
+impl BTreeMapCoinHelpers for BTreeMap<String, Uint128> {
+    fn into_vec(self) -> Vec<Coin> {
+        self.into_iter()
+            .map(|(denom, amount)| Coin { denom, amount })
+            .collect()
+    }
+
+    fn inplace_sub<'a, I>(&mut self, balance: I) -> StdResult<()>
+    where
+        I: IntoIterator<Item = (&'a String, &'a Uint128)>,
+    {
+        // Decrease total remaining supply
+        for (denom, amount) in balance {
+            if let Some(r) = self.get_mut(denom) {
+                if amount > r {
+                    return Err(StdError::generic_err(format!(
+                        "Subtract overflow for denom: {}",
+                        denom
+                    )));
+                }
+                *r -= amount;
+            } else {
+                return Err(StdError::generic_err(format!("Unknown denom {}", denom)));
+            }
+
+            // Remove denom if balance is now 0
+            if self.get(denom) == Some(&Uint128::zero()) {
+                self.remove(denom);
+            }
+        }
+        Ok(())
+    }
+
+    fn inplace_add<'a, I>(&mut self, balance: I) -> StdResult<()>
+    where
+        I: IntoIterator<Item = (&'a String, &'a Uint128)>,
+    {
+        for (denom, amount) in balance {
+            if let Some(counter) = self.get_mut(denom) {
+                *counter = counter.checked_add(*amount).map_err(|_| {
+                    StdError::generic_err(format!("Addition overflow for denom: {}", denom))
+                })?;
+            } else {
+                self.insert(denom.clone(), *amount);
+            }
+        }
+        Ok(())
+    }
+}
+
+pub trait VecCoinConversions {
+    fn to_tuple_iterator<'a>(&'a self) -> Box<dyn Iterator<Item = (&'a String, &'a Uint128)> + 'a>;
+    fn into_map(self) -> BTreeMap<String, Uint128>;
+    fn into_map_unique(self) -> StdResult<BTreeMap<String, Uint128>>;
+    fn to_formatted_string(&self) -> String;
+}
+
+impl VecCoinConversions for Vec<Coin> {
+    fn to_tuple_iterator<'a>(&'a self) -> Box<dyn Iterator<Item = (&'a String, &'a Uint128)> + 'a> {
+        Box::new(self.iter().map(|coin| (&coin.denom, &coin.amount)))
+    }
+
+    fn into_map(self) -> BTreeMap<String, Uint128> {
+        let mut denom_map = BTreeMap::new();
+
+        for coin in self {
+            if let Some(counter) = denom_map.get_mut(&coin.denom) {
+                *counter += coin.amount;
+            } else {
+                denom_map.insert(coin.denom, coin.amount);
+            }
+        }
+
+        denom_map
+    }
```

Comment:
Bellow code is just symbolic to show ahe idea (it is not Rust correct = it wonlt compile):
```suggestion
    fn into_map_onipresent(self, predicate: lambda[] (&mut self, balance Uint128)) -> BTreeMap<String, Uint128> {
        let mut denom_map = BTreeMap::new();

        for coin in self {
            if let Some(counter) = denom_map.get_mut(&coin.denom) {
                predicate(*counter, &coin.amount);
            } else {
                denom_map.insert(coin.denom, coin.amount);
            }
        }

        denom_map
    }
```